storage/interceptors/smaap_service_binding_interceptor.go

```diff
@@ -0,0 +1,521 @@
+			return f(ctx, obj)
+		}
+		operation, found := operations.GetFromContext(ctx)
+		if !found {
+			return nil, fmt.Errorf("operation missing from context")
+		}
+
+		osbClient, broker, service, plan, err := prepare(ctx, i.repository, i.osbClientCreateFunc, instance)
+		if err != nil {
+			return nil, err
+		}
+
+		if isBindable := !i.isPlanBindable(service, plan); isBindable {
+			return nil, &util.HTTPError{
+				ErrorType:   "BrokerError",
+				Description: fmt.Sprintf("plan %s is not bindable", plan.CatalogName),
+				StatusCode:  http.StatusBadGateway,
+			}
+		}
+
+		if isReady := i.isInstanceReady(instance); !isReady {
+			return nil, &util.HTTPError{
+				ErrorType:   "OperationInProgress",
+				Description: fmt.Sprintf("creation of instance %s is still in progress or failed", instance.Name),
+				StatusCode:  http.StatusUnprocessableEntity,
+			}
+		}
+
+		isDeleting, err := i.isInstanceDeleting(ctx, instance.ID)
+		if err != nil {
+			return nil, fmt.Errorf("could not determine instance state: %s", err)
+		}
+
+		if isDeleting {
+			return nil, &util.HTTPError{
+				ErrorType:   "OperationInProgress",
+				Description: fmt.Sprintf("instance %s is in state of delteion", instance.Name),
+				StatusCode:  http.StatusUnprocessableEntity,
+			}
+		}
+
+		var bindResponse *osbc.BindResponse
+		if !operation.Reschedule {
+			bindRequest := i.prepareBindRequest(instance, binding, service.CatalogID, plan.CatalogID)
+			contextBytes, err := json.Marshal(bindRequest.Context)
+			if err != nil {
+				return nil, fmt.Errorf("failed to marshal OSB context %+v: %s", bindRequest.Context, err)
+			}
+			binding.Context = contextBytes
+
+			log.C(ctx).Infof("Sending bind request %+v to broker with name %s", bindRequest, broker.Name)
+			bindResponse, err = osbClient.Bind(bindRequest)
+			if err != nil {
+				brokerError := &util.HTTPError{
+					ErrorType:   "BrokerError",
+					Description: fmt.Sprintf("Failed bind request %+v: %s", bindRequest, err),
+					StatusCode:  http.StatusBadGateway,
+				}
+				if shouldStartOrphanMitigation(err) {
+					// store the instance so that later on we can do orphan mitigation
+					_, err := f(ctx, obj)
+					if err != nil {
+						return nil, fmt.Errorf("broker error %s caused orphan mitigation which required storing the resource which failed with: %s", brokerError, err)
+					}
+
+					// mark the operation as deletion scheduled meaning orphan mitigation is required
+					operation.DeletionScheduled = time.Now()
+					operation.Reschedule = false
+					if _, err := i.repository.Update(ctx, operation, query.LabelChanges{}); err != nil {
+						return nil, fmt.Errorf("failed to update operation with id %s to schedule orphan mitigation after broker error %s: %s", operation.ID, brokerError, err)
+					}
+				}
+				return nil, brokerError
+			}
+
+			if err := i.enrichBindingWithBindingResponse(binding, bindResponse); err != nil {
```

Comment:
If binding is done asynchronously the binding response does not contain the credentials. After successful operation (the poll on line 205). GET request should be done in order to get the binding details.

Reply:
👍 Only if `"bindings_retrievable" :true`

Reply:
and if false do we trigger orphan mitigation ?

Reply:
so what im thinking - if bindings_retrievable is false, we will send accepts_incomplete=false to the broker on Bind request (this way we say that if the broker does not support get binding, we only support sync bind broker calls). If bindings_retrievable is true, we send accepts_incomplete=true and if the response is async and then our Get binding call fails, we schedule orphan mitigation. Any opinions ?